README.md
# RC-tracks

A printable race track systems for remote controlled cars.

<!-- vscode-markdown-toc -->

-   [At a glance](#Ataglance)
-   [Install and requirements](#Installandrequirements)
    -   [OpenSCAD](#OpenSCAD)
    -   [Source code](#Sourcecode)
        -   [Download the zip file](#Downloadthezipfile)
        -   [Get the code from the repository](#Getthecodefromtherepository)
    -   [Configuration](#Configuration)
-   [Render the parts](#Rendertheparts)
-   [Post-scripts](#Post-scripts)

<!-- vscode-markdown-toc-config
	numbering=false
	autoSave=true
	/vscode-markdown-toc-config -->
<!-- /vscode-markdown-toc -->

## <a name='Ataglance'></a>At a glance

The project contains designs for a modular race track system aimed at remote controlled cars. The parts are designed using a scripting language, processed by [OpenSCAD](https://openscad.org/about.html). There is no ready to use 3D model files in the source code. If you are interested in such files, please look at the [released versions](https://github.com/jsconan/rc-tracks/releases) which contains 3D model files built with default parameters. However, if you want to customize these parts, tailoring them with your measures, you should rather download the source code and then tweak the configuration files before generating your own 3D model files.

The project contains two different systems, each targeting a particular RC car scale:

-   [1/24 to 1/32 scale](scale-24/README.md)

    ![RC track](./doc/rctrack-24.jpg)

-   [1/64 to 1/76 scale](scale-64/README.md)

    ![RC track](./doc/rctrack-64.jpg)

Please refer to the readme file for each variant in order to get the proper documentation.

> **Note:** the scales are given as an estimate, not all scales have been tested. The `1/24 to 1/32` variant has been tested with `1/28` scale models, while the `1/64 to 1/76` variant has been tested with `1/76` models.

## <a name='Installandrequirements'></a>Install and requirements

To properly use the project you first need to do some installations steps.

### <a name='OpenSCAD'></a>OpenSCAD

First of all you need [OpenSCAD](https://openscad.org/) to be installed. To do so, please go to the [OpenSCAD download page](https://openscad.org/downloads.html), an get the version suitable for your operating system.

### <a name='Sourcecode'></a>Source code

The source code is hosted on a [Git](https://git-scm.com/) repository. To get it you can either download a [zip file](https://github.com/jsconan/rc-tracks/archive/refs/heads/main.zip), or clone the repository locally.

#### <a name='Downloadthezipfile'></a>Download the zip file

The source code can be downloaded from the [GitHub repository](https://github.com/jsconan/rc-tracks).

You can download it as zip file from this [link](https://github.com/jsconan/rc-tracks/archive/refs/heads/main.zip).

As the project is using a shared library, that is not supplied with the package, you will need to also download it.

Download the zip file from this [link](https://github.com/jsconan/camelSCAD/archive/refs/heads/main.zip).

Then extract its content inside the folder `lib/camelSCAD`. Please make sure the folder directly contains the library and not an intermediate folder like `lib/camelSCAD/camelSCAD-main`. If this is the case, please move the content one folder up and delete the extra folder.

Note: the easiest way is to call the `init.sh` script, that will take care of this for you.

#### <a name='Getthecodefromtherepository'></a>Get the code from the repository

A git tool is needed if you intend to get the source code from the git repository. You can download one either from the [main git website](https://git-scm.com/downloads), or from [GitHub](https://docs.github.com/en/github-cli).

Once you have your git tool installed, open a console window, select a project folder (create it if needed), then run the following commands:

```
git clone https://github.com/jsconan/rc-tracks.git
cd rc-tracks
git submodule init
git submodule update
```

The source code should have been downloaded, as well as the libraries.

Note: you can also use a graphical interface tool. In this case, please make it can also install the submodules.

Note 2: the easiest way is to call the `init.sh` script, that will take care of this for you.

### <a name='Configuration'></a>Configuration

The configuration for each system may be changed locally. To not collide with changes in the main implementation, the configuration files are not versioned. Instead, you will find `-dist` files that contain a copy of the typical configuration.

From a fresh copy of the project, you will need first to create these configuration files. The easiest way is to call the `init.sh` script, that will take care of this for you. The configuration files will also be checked and created if needed each time the `render.sh` script is called.

This initial creation of the configuration files could also be done manually. Here is how:

1. Go to the `config` folder
2. Copy any of the files having a `-dist` suffix in their name to a file without this suffix.

    _Example:_ copy `config-dist.scad` to `config.scad`

Please refer to the readme file for each variant in order to get the proper documentation.

## <a name='Rendertheparts'></a>Render the parts

In each system, the parts are all located into the `parts` folder. They can all be opened separately in [OpenSCAD](https://openscad.org/) and rendered.

However, it is possible to render all parts in batch. Each system proposes a script `render.sh` to take care of this.

This is a command-line tool, which accept several parameters. By default, it will render the preset selected in the config.

It is possible to select another preset, as well as set a different number of cells for the containers.

The parameter `-h` make it show the help.

If the slicer [PrusaSlicer](https://github.com/prusa3d/PrusaSlicer) is installed, running the command `render.sh -s` will produce the [G-code](https://en.wikipedia.org/wiki/G-code) from the rendered STL files and using the setting from the file `config/config.ini`.

Please refer to the readme file for each variant in order to get the proper documentation.

## <a name='Post-scripts'></a>Post-scripts

For each system, when using the built-in script to render/slice the models in batch, you can also add custom post-processing scripts. There is one for `render.sh` and one for `slice.sh`. They are respectively named `post-render.sh` and `post-slice.sh`.

As these script can be defined locally, they are not part of the versioned content. Instead, you can find sample files at `post-render-dist.sh` and `post-slice-dist.sh`. To use them, you need to:

-   copy the dist file to a local file, for example: `cp post-render-dist.sh post-render.sh`.
-   you need to make sure the file is executable: `chmod +x post-render.sh`.
-   finally, add you own commands to the copy.

Please refer to the readme file for each variant in order to get the proper documentation.


HISTORY.md
# rc-tracks history

## [Version 1.1.0](https://github.com/jsconan/rc-tracks/releases/tag/1.1.0)

**For the 1/64 scale variant.**

Add a low-level config option to render large-curve tiles full size instead of a split in halves, and add parts for both modes.
Add larger curves with ratios 3 and 4.

Move config options that were low-level to the constants.
Also, refactor the way curves are placed, unify the way sets of elements are built, and improve naming consistency.
Rework a bit the animated pictures.

## [Version 1.0.0](https://github.com/jsconan/rc-tracks/releases/tag/1.0.0)

Add a variant of the race track system for 1/64 to 1/76 scale RC cars.

![scale 64](https://github.com/jsconan/rc-tracks/raw/2635fe50cdf437d53924655f017a9943ff9a0d68/doc/rctrack-64.jpg)

This variant can be fully printed, barriers and ground included. It still comes with a modular approach, each element being able to be printed separately. Full tiles can also be printed all at once (barriers + ground).

Regarding the changes impacting the whole project:

-   The structure and the config have been unified, common config is available at the top level.
-   The script files also got unified, with support added for post-processing scripts.
-   Readme files are provided with details for each variant.

## [Version 0.5.0](https://github.com/jsconan/rc-tracks/releases/tag/0.5.0)

Finalize the extract of the race track system for 1/24 to 1/32 scale RC cars from the repository [jsconan/things](https://github.com/jsconan/things).

-   Fix a typo in some comments.
-   Use the shared internal shape `linkProfile` instead of redefining it in `barrierLinkProfile`.
-   Add a default printer config for the slicer.
-   Add a slicer script.

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/59

## [Version 0.4.2](https://github.com/jsconan/rc-tracks/releases/tag/0.4.2)

Fixed design of the race track system for 1/24 to 1/32 scale RC cars.

-   Fix the length of the special barrier body related to a U-turn compensation barrier.

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/49

## [Version 0.4.1](https://github.com/jsconan/rc-tracks/releases/tag/0.4.1)

Fixed design of the race track system for 1/24 to 1/32 scale RC cars.

-   Fix a computation issue on the number of notches for the straight barrier holders (when the length is a multiple of the normal size)
-   Improve the design of the U-turn barrier holders by adding a pocket to prevent round edges on the slot near the tower
-   Dispatch the "special" shapes into more appropriate files (arch and connector)

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/48

## [Version 0.4.0](https://github.com/jsconan/rc-tracks/releases/tag/0.4.0)

Improved design of the race track system for 1/24 to 1/32 scale RC cars.

-   Encapsulate the shapes into final modules
-   Add sizing functions
-   Add more printer constraints (size of the printer's plate)
-   Add print plates for sets of elements
-   The file structure has been updated.
-   Add a special barrier connector to link barrier holders to unibody barriers (straight version only)
-   The render script has been reworked, adding more options

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/47

## [Version 0.3.0](https://github.com/jsconan/rc-tracks/releases/tag/0.3.0)

Improved design of the race track system for 1/24 to 1/32 scale RC cars.

-   The file structure has been updated.
-   A new kind of barrier has been added, defining a unibody variant: each track element is built from a single piece instead of 3.
-   The render script has been reworked, adding more options
-   The track can now have curves that are not directly related to the width lanes
-   Add wavy flags in the set of accessories

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/46

## [Version 0.2.0](https://github.com/jsconan/rc-tracks/releases/tag/0.2.0)

Improved design of the race track system for 1/24 to 1/32 scale RC cars.

-   The file structure has been updated.
-   The link between parts has been reworked for a better track cohesion.
-   The profiles have been redesigned.
-   A distinction is made between the inner and outer curves and the parts can be resized to a multiple of the main length.
-   Added shapes:
    -   Straight track chunks: simple and double length
    -   Curved track chunks: inner, outer, small, short, U-turn
    -   Special track chunks: arch to constraint the track width between two sides, U-turn compensation
    -   Track samples to draw a model of the track (default scaled to 1/10)
    -   Track accessories: led-clip, mast, flag

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/40

## [Version 0.1.0](https://github.com/jsconan/rc-tracks/releases/tag/0.1.0)

Design a race track system for 1/24 to 1/32 scale RC cars.

---

Notes:

-   Import from the repository [jsconan/things](https://github.com/jsconan/things)
-   Extract of the pull request https://github.com/jsconan/things/pull/37


scale-24/render.sh
#!/bin/bash
#
# GPLv3 License
#
# Copyright (c) 2020-2022 Jean-Sebastien CONAN
#
# This file is part of jsconan/rc-tracks.
#
# jsconan/rc-tracks is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# jsconan/rc-tracks is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with jsconan/rc-tracks. If not, see <http://www.gnu.org/licenses/>.
#

#
# A race track system for 1/24 to 1/32 scale RC cars.
#
# Generates the STL files for the project.
#
# @author jsconan
#

# application params
trackSectionLength=
trackSectionWidth=
trackLaneWidth=
trackRadius=
barrierHeight=
sampleSize=

# script config
scriptpath="$(dirname $0)"
project="$(pwd)"
srcpath="${project}"
dstpath="${project}/dist/stl"
slcpath="${project}/dist/gcode"
configpath="${srcpath}/config"
partpath="${srcpath}/parts"
platepath="${srcpath}/plates"
format=
parallel=
showConfig=
renderAccessories=
renderElements=
renderUnibody=
renderSamples=
renderPlates=
cleanUp=
slice=

# include libs
source "${scriptpath}/../lib/camelSCAD/scripts/utils.sh"

# Builds the list of config parameters.
# @param right - Right oriented or left oriented
paramlist() {
    local rightOriented=$1
    local params=(
        "$(varif "trackSectionLength" ${trackSectionLength})" \
        "$(varif "trackSectionWidth" ${trackSectionWidth})" \
        "$(varif "trackLaneWidth" ${trackLaneWidth})" \
        "$(varif "trackRadius" ${trackRadius})" \
        "$(varif "barrierHeight" ${barrierHeight})" \
        "$(varif "sampleSize" ${sampleSize})" \
        "$(varif "rightOriented" ${rightOriented})"
    )
    echo "${params[@]}"
}

# Renders the files from a path.
#
# @param sourcepath - The path of the folder containing the SCAD files to render.
# @param destpath - The path to the output folder.
# @param right - Right oriented or left oriented
# @param prefix - Optional prefix added to the output fil name
# @param suffix - Optional suffix added to the output fil name
renderpath() {
    scadrenderall "$1" "$2" "$4" "$5" --quiet $(paramlist $3)
}

# Renders the files from a path, taking care of the curves.
#
# @param sourcepath - The path of the folder containing the SCAD files to render.
# @param destpath - The path to the output folder.
renderpathall() {
    printmessage "${C_MSG}- straight elements"
    renderpath "$1/straight" "$2"
    printmessage "${C_MSG}- left curved elements"
    renderpath "$1/curved" "$2" "0" "left"
    printmessage "${C_MSG}- right curved elements"
    renderpath "$1/curved" "$2" "1" "right"
}

# Display the render config
showconfig() {
    local input="${configpath}/print.scad"
    local output="${dstpath}/print.echo"
    local config="${dstpath}/config.txt"
    createpath "${dstpath}" "output"
    printmessage "${C_MSG}The track elements would be generated with respect to the following config:"
    scadecho "${input}" "${dstpath}" "" "" $(paramlist) > /dev/null
    sed '1d; $d' "${output}" > "${config}"
    rm "${output}" > /dev/null
    cat "${config}"
}

# load parameters
while (( "$#" )); do
    case $1 in
        "a"|"accessories")
            renderAccessories=1
            showConfig=1
        ;;
        "e"|"elements")
            renderElements=1
            showConfig=1
        ;;
        "u"|"unibody")
            renderUnibody=1
            showConfig=1
        ;;
        "s"|"samples")
            renderSamples=1
            showConfig=1
        ;;
        "p"|"plates")
            renderPlates=1
        ;;
        "c"|"config")
            showConfig=1
        ;;
        "-l"|"--length")
            trackSectionLength=$2
            shift
        ;;
        "-w"|"--width")
            trackSectionWidth=$2
            shift
        ;;
        "-b"|"--height")
            barrierHeight=$2
            shift
        ;;
        "-r"|"--radius")
            trackRadius=$2
            shift
        ;;
        "-t"|"--track")
            trackLaneWidth=$2
            shift
        ;;
        "-m"|"--sample")
            sampleSize=$2
            shift
        ;;
        "-f"|"--format")
            format=$2
            shift
        ;;
        "-p"|"--parallel")
            parallel=$2
            shift
        ;;
        "-s"|"--slice")
            slice=1
        ;;
        "-c"|"--clean")
            cleanUp=1
        ;;
        "-h"|"--help")
            echo -e "${C_INF}Renders OpenSCAD files${C_RST}"
            echo -e "  ${C_INF}Usage:${C_RST}"
            echo -e "${C_CTX}\t$0 [command] [-h|--help] [-o|--option value] files${C_RST}"
            echo
            echo -e "${C_MSG}  a,  accessories     ${C_RST}Render the accessories"
            echo -e "${C_MSG}  e,  elements        ${C_RST}Render the track separated elements"
            echo -e "${C_MSG}  u,  unibody         ${C_RST}Render the track unibody elements"
            echo -e "${C_MSG}  s,  samples         ${C_RST}Render the samples"
            echo -e "${C_MSG}  p,  plates          ${C_RST}Render the plates"
            echo -e "${C_MSG}  c,  config          ${C_RST}Show the config values"
            echo -e "${C_MSG}  -h,  --help         ${C_RST}Show this help"
            echo -e "${C_MSG}  -l,  --length       ${C_RST}Set the length of a track section"
            echo -e "${C_MSG}  -w,  --width        ${C_RST}Set the virtual width of a track lane (used to compute the radius)"
            echo -e "${C_MSG}  -t   --track        ${C_RST}Set the actual width of a track lane (physical width, used for the arches)"
            echo -e "${C_MSG}  -b   --height       ${C_RST}Set the height of the track barrier"
            echo -e "${C_MSG}  -r   --radius       ${C_RST}Set the radius of the track inner curve"
            echo -e "${C_MSG}  -m   --sample       ${C_RST}Set the size of sample element"
            echo -e "${C_MSG}  -f   --format       ${C_RST}Set the output format"
            echo -e "${C_MSG}  -p   --parallel     ${C_RST}Set the number of parallel processes"
            echo -e "${C_MSG}  -s   --slice        ${C_RST}Slice the rendered files using the default configuration"
            echo -e "${C_MSG}  -c   --clean        ${C_RST}Clean up the output folder before rendering"
            echo
            exit 0
        ;;
        *)
            ls $1 >/dev/null 2>&1
            if [ "$?" == "0" ]; then
                srcpath=$1
            else
                printerror "Unknown parameter ${1}"
            fi
        ;;
    esac
    shift
done

# align values
if [ "${trackSectionLength}" != "" ]; then
    if [ "${trackSectionWidth}" == "" ]; then
        trackSectionWidth=$((${trackSectionLength} * 2))
    fi
    if [ "${trackLaneWidth}" == "" ]; then
        trackLaneWidth=$((${trackSectionLength} * 3))
    fi
    if [ "${trackRadius}" == "" ]; then
        trackRadius=${trackSectionLength}
    fi
fi

# default script config
if [ "${renderAccessories}" == "" ] && [ "${renderElements}" == "" ] && [ "${renderUnibody}" == "" ] && [ "${renderSamples}" == "" ] && [ "${showConfig}" == "" ]; then
    renderAccessories=1
    renderElements=1
    renderUnibody=1
    renderSamples=1
    showConfig=1
fi

# check OpenSCAD
scadcheck

# defines the output format
scadformat "${format}"

# defines the number of parallel processes
scadprocesses "${parallel}"

# clean up the output
if [ "${cleanUp}" != "" ]; then
    printmessage "${C_CTX}Cleaning up the output folder"
    rm -rf "${dstpath}"

    if [ "${slice}" != "" ]; then
        printmessage "${C_CTX}Cleaning up the slicer output folder"
        rm -rf "${slcpath}"
    fi
fi

# make sure the config exists
distfile "${configpath}/config.scad"

# show the config
if [ "${showConfig}" != "" ]; then
    showconfig
fi

# render the plates
if [ "${renderPlates}" != "" ]; then
    printmessage "${C_CTX}Will render the print plates"
    partpath=${platepath}
    dstpath=${dstpath}/plates
else
    printmessage "${C_CTX}Will render the parts separately"
fi

# render the files
if [ "${renderAccessories}" != "" ]; then
    printmessage "${C_MSG}Rendering accessories"
    renderpath "${partpath}/accessories" "${dstpath}/accessories"
fi
if [ "${renderElements}" != "" ]; then
    printmessage "${C_MSG}Rendering track separated elements"
    renderpathall "${partpath}/elements" "${dstpath}/elements"
fi
if [ "${renderUnibody}" != "" ]; then
    printmessage "${C_MSG}Rendering track unibody elements"
    renderpathall "${partpath}/unibody" "${dstpath}/unibody"
fi
if [ "${renderSamples}" != "" ]; then
    printmessage "${C_MSG}Rendering track samples"
    renderpathall "${partpath}/samples" "${dstpath}/samples"
fi

# run a post-render script
if [ -x "${scriptpath}/post-render.sh" ]; then
    printmessage "${C_CTX}Calling the post-render script"
    "${scriptpath}/post-render.sh"
fi

# slice the rendered files
if [ "${slice}" != "" ]; then
    printmessage "${C_CTX}Slicing the rendered files"
    ./slice.sh
fi
